### PR TITLE
server: remove error stack in log file when parsing failed

### PR DESCRIPTION
**Pull request by peterepsteen**
Remove useless error stacks in logs for parseing errors. Solves issue #15017.

### What problem does this PR solve? 
Adds a function that will determine whether or not to skip logging an error's stack trace after
a dispatched command fails. For now, it only will skip if the error is a parse error, but 
more errors can be added to skip as needed.

### What is changed and how it works?
Add check before logging a failed command's full stack trace:
```

doLogStackTrace := !shouldSkipStackTrace(
if doLogStackTrace {
...
```
```
func shouldSkipStackTrace(err error) bool {
	return parser.ErrParse.Equal(err)
}
```

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)
Before: 
```
[2020/02/29 11:39:03.030 -07:00] [WARN] [conn.go:698] ["command dispatched failed"] [conn=1] [connInfo="id:1, addr:127.0.0.1:47146 status:10, collation:utf8_general_ci, user:root"] [command=Query] [status="inTxn:0, autocommit:1"] [sql="show tablsaa"] [txn_mode=OPTIMISTIC] [err="[parser:1064]You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 12 near \"tablsaa\" \ngithub.com/pingcap/errors.AddStack\n\t/home/peter/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20190809092503-95897b64e011/errors.go:174\ngithub.com/pingcap/parser
...
```
After: 
```
[2020/02/29 12:54:31.122 -07:00] [WARN] [session.go:1095] ["parse SQL failed"] [conn=1] [error="line 1 column 12 near \"tablsaa\" "] [SQL="show tablsaa"]
[2020/02/29 12:54:31.122 -07:00] [ERROR] [misc.go:121] ["syntax error"] [error="line 1 column 12 near \"tablsaa\" "]
```

Code changes

Side effects

Related changes

Release note

 - Don't log error stack when parsing failed

PS: First time contributing to this project so let me know if I am missing anything. Thanks!

### CODE REVIEW

File: server/conn.go [hunk 1]
```diff
@@ -691,18 +691,23 @@ func (cc *clientConn) Run(ctx context.Context) {
 				}
 				return
 			}
-			var txnMode string
-			if cc.ctx != nil {
-				txnMode = cc.ctx.GetSessionVars().GetReadableTxnMode()
+
+			doLogStackTrace := !shouldSkipStackTrace(err)
+			if doLogStackTrace {
```
Comment (zz-jason) [inline server/conn.go]: We should do this check in `errStrForLog()`.

**Comment by zz-jason** [comment]
BTW, I think this PR is not to fix #15018, it fixes #15017 actually 😂

**Comment by peterepsteen** [comment]
> BTW, I think this PR is not to fix #15018, it fixes #15017 actually 

My bad! Edited.

File: server/conn.go (hunk 1 above)
Comment (peterepsteen) [reply server/conn.go]: Ah didn't see that. Thanks. Just changed it.

File: server/conn.go [hunk 2]
```diff
@@ -739,6 +739,8 @@ func errStrForLog(err error) string {
 	if kv.ErrKeyExists.Equal(err) {
 		// Do not log stack for duplicated entry error.
 		return err.Error()
+	} else if parser.ErrParse.Equal(err) {
+		return err.Error()
 	}
```
Comment (Deardrops) [inline server/conn.go]:
```suggestion
	if kv.ErrKeyExists.Equal(err) || parser.ErrParse.Equal(err) {
		// Do not log stack for duplicated entry error and parser error.
		return err.Error()
	}
```

**Comment by peterepsteen** [comment]
Squashed the commits into one so the message is more relevant

**Comment by zz-jason** [comment]
Hi @peterepsteen, this is an issue for the challenge program(Usability Challenge Program). congratulations that you are almost finished it! would you like to join this event? You can sign up by [opening an issue](https://github.com/tidb-challenge-program/register/issues/new?assignees=&labels=&template=challenge-program-register.md&title=UCP%2FSignup) in [tidb-challenge-program/register](https://github.com/tidb-challenge-program/register) repo.

As the [issue body](https://github.com/pingcap/tidb/issues/15017) said, after finishing this issue, you can gain 50 score. After gaining 200 score by finishing `diffucilty/easy` issues, you can challenge the `diffuculty/medium` or `diffucilty/hard` issue listed in the [challenge-program](https://github.com/pingcap/tidb/projects/26) project.

**Pull request merged**